<a href="https://colab.research.google.com/github/drshahizan/Python-big-data/blob/main/assignment/ass6/bdm/Homelander/Homelander_BDM_Assignment6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

###Homelander: Assignment 6
1. Pang Chern Hong  |MCS231006

2. Nian Cong      |MCS231007

In [1]:

# Upload kaggle.json API token, and download / unzip Flight Delay and Cancellation Dataset

# Install and upload the kaggle.json file
!pip install kaggle

from google.colab import files
files.upload()

!mkdir ~/.kaggle
!cp kaggle.json ~/.kaggle/
!chmod 600 ~/.kaggle/kaggle.json
!kaggle datasets download -d patrickzel/flight-delay-and-cancellation-dataset-2019-2023
!unzip flight-delay-and-cancellation-dataset-2019-2023.zip


Saving kaggle.json to kaggle.json
 96% 134M/140M [00:02<00:00, 46.1MB/s]
100% 140M/140M [00:02<00:00, 68.5MB/s]
Archive:  flight-delay-and-cancellation-dataset-2019-2023.zip
  inflating: dictionary.html         
  inflating: flights_sample_3m.csv   


In [2]:
#Install modin
!pip install modin[all]

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 5.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.3/12.3 MB 37.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.6/62.6 MB 6.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 54.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 341.8/341.8 kB 27.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 201.4/201.4 kB 16.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 63.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.1/98.1 kB 11.7 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 128.2/128.2 kB 11.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.7/8.7 MB 53.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.4/58.4 kB 

In [4]:
import numpy as np
import time
import pandas as pd
import modin.pandas as md

## Example

In [5]:
#Create an array of random integers
arr =np.random.randint(low=10,high=1000,size=(2**15,2**10))
#Save it as a csv file
np.savetxt("data.csv", arr, delimiter=",")

In [6]:
%time p_df = pd.read_csv("data.csv")

CPU times: user 13.3 s, sys: 4.1 s, total: 17.4 s
Wall time: 17.9 s


In [7]:
#Read data.csv file using Modin
%time m_df = md.read_csv("data.csv")

To remove this warning, run the following python code before doing dataframe operations:

    import ray
    ray.init()

2023-12-17 04:07:57,475	INFO worker.py:1673 -- Started a local Ray instance.


CPU times: user 2.96 s, sys: 1.97 s, total: 4.93 s
Wall time: 30.5 s


## Data loading

## Method 1: Load Less Data

In [8]:
!pip install pandas
!pip install dask
!pip install "dask[complete]"
!pip install vaex
!pip install modin

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 6.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.5/4.5 MB 18.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.2/43.2 kB 4.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.2/58.2 kB 6.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 46.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 64.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 511.6/511.6 kB 42.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.7/11.7 MB 33.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 93.1/93.1 kB 9.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 59.7/59.7 kB 6.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 260.7/260.7 kB 25.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.7/2.7 MB 60.7 MB/s eta 0:00:0

In [12]:
%%time
with open('flights_sample_3m.csv') as file:
    n_rows = len(file.readlines())

print ('Exact number of rows: {}'.format(n_rows))

Exact number of rows: 3000001
CPU times: user 1.06 s, sys: 1.76 s, total: 2.82 s
Wall time: 4.57 s


In [17]:
%%time
df = pd.read_csv("flights_sample_3m.csv")

CPU times: user 18.2 s, sys: 3.92 s, total: 22.1 s
Wall time: 23 s


In [18]:
df.info(memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000000 entries, 0 to 2999999
Data columns (total 32 columns):
 #   Column                   Dtype  
---  ------                   -----  
 0   FL_DATE                  object 
 1   AIRLINE                  object 
 2   AIRLINE_DOT              object 
 3   AIRLINE_CODE             object 
 4   DOT_CODE                 int64  
 5   FL_NUMBER                int64  
 6   ORIGIN                   object 
 7   ORIGIN_CITY              object 
 8   DEST                     object 
 9   DEST_CITY                object 
 10  CRS_DEP_TIME             int64  
 11  DEP_TIME                 float64
 12  DEP_DELAY                float64
 13  TAXI_OUT                 float64
 14  WHEELS_OFF               float64
 15  WHEELS_ON                float64
 16  TAXI_IN                  float64
 17  CRS_ARR_TIME             int64  
 18  ARR_TIME                 float64
 19  ARR_DELAY                float64
 20  CANCELLED                float64
 21  CANCELLA

In [19]:
df.info(memory_usage=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000000 entries, 0 to 2999999
Data columns (total 32 columns):
 #   Column                   Dtype  
---  ------                   -----  
 0   FL_DATE                  object 
 1   AIRLINE                  object 
 2   AIRLINE_DOT              object 
 3   AIRLINE_CODE             object 
 4   DOT_CODE                 int64  
 5   FL_NUMBER                int64  
 6   ORIGIN                   object 
 7   ORIGIN_CITY              object 
 8   DEST                     object 
 9   DEST_CITY                object 
 10  CRS_DEP_TIME             int64  
 11  DEP_TIME                 float64
 12  DEP_DELAY                float64
 13  TAXI_OUT                 float64
 14  WHEELS_OFF               float64
 15  WHEELS_ON                float64
 16  TAXI_IN                  float64
 17  CRS_ARR_TIME             int64  
 18  ARR_TIME                 float64
 19  ARR_DELAY                float64
 20  CANCELLED                float64
 21  CANCELLA

In [21]:
import psutil
# Record start time
start_time = time.time()

# Load games.csv without 'mac' and 'linux' columns
df_load = pd.read_csv('flights_sample_3m.csv', usecols=lambda col: col not in ['AIRLINE_DOT'])

# Record end time
end_time = time.time()

# Display DataFrame info
print("DataFrame 'df_load' Info:")
print(df_load.info())

# Display system RAM usage as a percentage
ram_percentage = (psutil.virtual_memory().used / psutil.virtual_memory().total) * 100
print("\nSystem RAM Usage:")
print(f"Used RAM: {ram_percentage:.2f}%")

# Display DataFrame memory usage as a percentage
df_memory_percentage = (df_load.memory_usage().sum() / psutil.virtual_memory().total) * 100
print("\nDataFrame Memory Usage:")
print(f"Used Memory: {df_memory_percentage:.2f}%")

# Display processing time
print("\nProcessing Time:")
print(f"Time taken: {end_time - start_time:.2f} seconds")

DataFrame 'df_load' Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000000 entries, 0 to 2999999
Data columns (total 31 columns):
 #   Column                   Dtype  
---  ------                   -----  
 0   FL_DATE                  object 
 1   AIRLINE                  object 
 2   AIRLINE_CODE             object 
 3   DOT_CODE                 int64  
 4   FL_NUMBER                int64  
 5   ORIGIN                   object 
 6   ORIGIN_CITY              object 
 7   DEST                     object 
 8   DEST_CITY                object 
 9   CRS_DEP_TIME             int64  
 10  DEP_TIME                 float64
 11  DEP_DELAY                float64
 12  TAXI_OUT                 float64
 13  WHEELS_OFF               float64
 14  WHEELS_ON                float64
 15  TAXI_IN                  float64
 16  CRS_ARR_TIME             int64  
 17  ARR_TIME                 float64
 18  ARR_DELAY                float64
 19  CANCELLED                float64
 20  CANCELLATION_CODE   

Compare to the full dataset loading, if we drop one of the column named Airline_dot, we found that the memory usage dropped from 732.4MB to 709.5MB, if we drop more unnecessary columns, we can save more memory and also the processing time needed.

In [23]:
%%time
# Read in the CSV into a Pandas DataFrame for first 10% of total rows

df = pd.read_csv('flights_sample_3m.csv', nrows=round(0.1*n_rows))
df = df.dropna(axis=1)

print(df)
df.info(memory_usage=True)

           FL_DATE                 AIRLINE                 AIRLINE_DOT  \
0       2019-01-09   United Air Lines Inc.   United Air Lines Inc.: UA   
1       2022-11-19    Delta Air Lines Inc.    Delta Air Lines Inc.: DL   
2       2022-07-22   United Air Lines Inc.   United Air Lines Inc.: UA   
3       2023-03-06    Delta Air Lines Inc.    Delta Air Lines Inc.: DL   
4       2020-02-23        Spirit Air Lines        Spirit Air Lines: NK   
...            ...                     ...                         ...   
299995  2023-07-28         JetBlue Airways         JetBlue Airways: B6   
299996  2020-11-04  Southwest Airlines Co.  Southwest Airlines Co.: WN   
299997  2021-09-28  Southwest Airlines Co.  Southwest Airlines Co.: WN   
299998  2022-03-23             Horizon Air             Horizon Air: QX   
299999  2021-04-04   United Air Lines Inc.   United Air Lines Inc.: UA   

       AIRLINE_CODE  DOT_CODE  FL_NUMBER ORIGIN          ORIGIN_CITY DEST  \
0                UA     19977     

If we load only 10% of the data, we now only used 34.3MB memory and CPU times of 1.85 s.

###USE CHUNKING

Chunking is a technique that divide the dataset into smaller chunks, then after these chunks being processed, they combine to form the results.

In [27]:
## check whether there is null values
df.isnull().sum()

FL_DATE         0
AIRLINE         0
AIRLINE_DOT     0
AIRLINE_CODE    0
DOT_CODE        0
FL_NUMBER       0
ORIGIN          0
ORIGIN_CITY     0
DEST            0
DEST_CITY       0
CRS_DEP_TIME    0
CRS_ARR_TIME    0
CANCELLED       0
DIVERTED        0
DISTANCE        0
dtype: int64

In [28]:
## Check whether there are any missing values
df.isna().sum()

FL_DATE         0
AIRLINE         0
AIRLINE_DOT     0
AIRLINE_CODE    0
DOT_CODE        0
FL_NUMBER       0
ORIGIN          0
ORIGIN_CITY     0
DEST            0
DEST_CITY       0
CRS_DEP_TIME    0
CRS_ARR_TIME    0
CANCELLED       0
DIVERTED        0
DISTANCE        0
dtype: int64

Anyway, we just check again all these data to make sure they have no missing values, if there are any, for these numerical values, we fill it with the mean of each chunk.

In [35]:
# Measure memory usage before
memory_before = psutil.Process().memory_info().rss / (1024 * 1024)  # in megabytes

# Measure time before
start_time = time.time()

chunk_size = 100000

# Create an empty DataFrame to store the processed chunks
processed_chunks = []

# Iterate over chunks
for chunk in pd.read_csv('flights_sample_3m.csv', chunksize=chunk_size):
    # Remove duplicate rows
    chunk = chunk.drop_duplicates()

    # Remove empty rows (assuming empty rows have NaN values in all columns)
    chunk = chunk.dropna(how='all')

    chunk.columns = ['FL_DATE', 'AIRLINE', "AIRLINE_DOT", 'AIRLINE_CODE', 'DOT_CODE', 'FL_NUMBER', 'ORIGIN', 'ORIGIN_CITY',
                      'DEST', 'DEST_CITY', 'CRS_DEP_TIME', 'DEP_TIME', 'DEP_DELAY', 'TAXI_OUT',
                      'WHEELS_OFF', 'WHEELS_ON', 'TAXI_IN', 'CRS_ARR_TIME', 'ARR_TIME', 'ARR_DELAY', 'CANCELLED',
                      'CANCELLATION_CODE', 'DIVERTED', 'CRS_ELAPSED_TIME', 'ELAPSED_TIME', 'AIR_TIME', "DISTANCE",
                      "DELAY_DUE_CARRIER", "DELAY_DUE_WEATHER", "DELAY_DUE_NAS", "DELAY_DUE_SECURITY","DELAY_DUE_LATE_AIRCRAFT"]


    # Impute Numerical Values
    chunk['DEP_DELAY'].fillna(chunk['DEP_DELAY'].mean(), inplace=True)
    chunk['CRS_ARR_TIME'].fillna(chunk['CRS_ARR_TIME'].mean(), inplace=True)
    chunk['CRS_ELAPSED_TIME'].fillna(chunk['CRS_ELAPSED_TIME'].mean(), inplace=True)
    chunk['ELAPSED_TIME'].fillna(chunk['ELAPSED_TIME'].mean(), inplace=True)
    chunk['DISTANCE'].fillna(chunk['DISTANCE'].mean(), inplace=True)
    chunk['CRS_DEP_TIME'].fillna(chunk['CRS_DEP_TIME'].mean(), inplace=True)


    # Append the processed chunk to the list
    processed_chunks.append(chunk)

# Concatenate the processed chunks into a final DataFrame
df_cleaned = pd.concat(processed_chunks, ignore_index=True)

# Measure time after
end_time = time.time()

# Measure memory usage after
memory_after = psutil.Process().memory_info().rss / (1024 * 1024)  # in megabytes

# Calculate differences
execution_time = end_time - start_time
memory_difference = memory_after - memory_before

# Print results
print(f"Execution Time: {execution_time} seconds")
print(f"Memory Usage Before: {memory_before} MB")
print(f"Memory Usage After: {memory_after} MB")
print(f"Memory Usage Difference: {memory_difference} MB")


Execution Time: 24.233595609664917 seconds
Memory Usage Before: 1911.21484375 MB
Memory Usage After: 3297.1171875 MB
Memory Usage Difference: 1385.90234375 MB


##OPTIMIZE DATA TYPES

In [70]:
# Make copies of the original DataFrames
start_time = time.time()
df_copy = df.copy()

# Convert columns in df_game_copy
df_copy['CANCELLED'] = df_copy['CANCELLED'].astype('bool')
df_copy['DIVERTED'] = df_copy['DIVERTED'].astype('bool')
df_copy['CRS_ARR_TIME'] = df_copy['CRS_ARR_TIME'].astype('uint32')
df_copy['CRS_DEP_TIME'] = df_copy['CRS_DEP_TIME'].astype('uint32')
df_copy['FL_NUMBER'] = df_copy['FL_NUMBER'].astype('uint32')
df_copy['DOT_CODE'] = df_copy['DOT_CODE'].astype('uint32')
df_copy['DEP_TIME'] = df_copy['DOT_CODE'].astype('float32')
df_copy['DEP_DELAY'] = df_copy['DEP_DELAY'].astype('float32')
df_copy['TAXI_OUT'] = df_copy['TAXI_OUT'].astype('float32')
df_copy['WHEELS_OFF'] = df_copy['WHEELS_OFF'].astype('float32')
df_copy['WHEELS_ON'] = df_copy['WHEELS_ON'].astype('float32')
df_copy['TAXI_IN'] = df_copy['TAXI_IN'].astype('float32')
df_copy['ARR_TIME'] = df_copy['ARR_TIME'].astype('float32')
df_copy['ARR_DELAY'] = df_copy['ARR_DELAY'].astype('float32')
df_copy['CRS_ELAPSED_TIME'] = df_copy['ELAPSED_TIME'].astype('float32')
df_copy['ELAPSED_TIME'] = df_copy['ELAPSED_TIME'].astype('float32')
df_copy['AIR_TIME'] = df_copy['AIR_TIME'].astype('float32')
df_copy['DISTANCE'] = df_copy['DISTANCE'].astype('float32')
df_copy['DELAY_DUE_CARRIER'] = df_copy['DELAY_DUE_CARRIER'].astype('float32')
df_copy['DELAY_DUE_WEATHER'] = df_copy['DELAY_DUE_WEATHER'].astype('float32')
df_copy['DELAY_DUE_NAS'] = df_copy['DELAY_DUE_NAS'].astype('float32')
df_copy['DELAY_DUE_SECURITY'] = df_copy['DELAY_DUE_SECURITY'].astype('float32')
df_copy['DELAY_DUE_LATE_AIRCRAFT'] = df_copy['DELAY_DUE_NAS'].astype('float32')

In [71]:
# Display system computing RAM (in percentage) after converting columns
ram_percentage_after_conversion = psutil.virtual_memory().percent
print(f"System RAM Usage After Conversion: {ram_percentage_after_conversion:.2f}%")

# Calculate processing time in seconds
end_time = time.time()
total_time_seconds = end_time - start_time
print(f"Processing Time: {total_time_seconds:.2f} seconds")

# Check the data size after conversion
df_copy.info(memory_usage='True')

System RAM Usage After Conversion: 60.80%
Processing Time: 2.32 seconds
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000000 entries, 0 to 2999999
Data columns (total 32 columns):
 #   Column                   Dtype  
---  ------                   -----  
 0   FL_DATE                  object 
 1   AIRLINE                  object 
 2   AIRLINE_DOT              object 
 3   AIRLINE_CODE             object 
 4   DOT_CODE                 uint32 
 5   FL_NUMBER                uint32 
 6   ORIGIN                   object 
 7   ORIGIN_CITY              object 
 8   DEST                     object 
 9   DEST_CITY                object 
 10  CRS_DEP_TIME             uint32 
 11  DEP_TIME                 float32
 12  DEP_DELAY                float32
 13  TAXI_OUT                 float32
 14  WHEELS_OFF               float32
 15  WHEELS_ON                float32
 16  TAXI_IN                  float32
 17  CRS_ARR_TIME             uint32 
 18  ARR_TIME                 float32
 19  ARR_DELAY   

Originally the 10% of dataset need 658 MB to load, but after we convert the data from float to bool for columns CANCELLED and DIVERTED, and convert the data types of int64 to uint32, we greatly saved the memory for CPU and also the run times.

##SAMPLING

Now we load the data of 10% but this time the data is more distributed, so this 10% of data can be more representative compared to only the first 10%。

In [51]:
import pandas as pd
import numpy as np
import time
import psutil

df = pd.read_csv("flights_sample_3m.csv")

# Record start time
start_time = time.time()

# Perform sampling on each DataFrame
sample = df.sample(frac=0.1, random_state=np.random.randint(10))

end_time = time.time()
total_time_seconds = end_time - start_time
print(f"Processing Time: {total_time_seconds:.2f} seconds")

# Get RAM usage percentage
ram_percentage = psutil.virtual_memory().percent
print(f"System RAM Usage: {ram_percentage:.2f}%")

# Check the final_combined data size
sample.info(memory_usage='TRUE')

Processing Time: 0.98 seconds
System RAM Usage: 49.80%
<class 'pandas.core.frame.DataFrame'>
Index: 300000 entries, 2288504 to 267028
Data columns (total 32 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   FL_DATE                  300000 non-null  object 
 1   AIRLINE                  300000 non-null  object 
 2   AIRLINE_DOT              300000 non-null  object 
 3   AIRLINE_CODE             300000 non-null  object 
 4   DOT_CODE                 300000 non-null  int64  
 5   FL_NUMBER                300000 non-null  int64  
 6   ORIGIN                   300000 non-null  object 
 7   ORIGIN_CITY              300000 non-null  object 
 8   DEST                     300000 non-null  object 
 9   DEST_CITY                300000 non-null  object 
 10  CRS_DEP_TIME             300000 non-null  int64  
 11  DEP_TIME                 292239 non-null  float64
 12  DEP_DELAY                292237 non-null  float64
 13  TAX

This is actually comply with the result that loading the full data set need around 750MB of memory, while 10% of data need 75.5 MB memory, by properly sampling the data, we can greatly save our time and memory usage without damaging the reprentability of the data.

##PARALLELIZE WITH DASK

Modern Laptop mostly has 8 cores, while the pandas can only use 1 of the cores to process and read the data, this make the CPU unable to fully function and its efficiency will be definitely dragged down especially when the dataset is large. Hence, Dask is used to process these data parallelly.



In [54]:
%time
import dask.dataframe as dd

ddf = dd.from_pandas(df, npartitions=4)

CPU times: user 4 µs, sys: 1e+03 ns, total: 5 µs
Wall time: 9.3 µs


Now we just calculate a parameter to see the processing time with dask.

In [57]:
# Measure memory usage before
memory_before = psutil.Process().memory_info().rss / (1024 * 1024)  # in megabytes
# Measure time before
start_time = time.time()
ddf = dd.from_pandas(df, npartitions=4)

# Now, perform parallelized operations
mean_time = ddf['CRS_ARR_TIME'].mean()

# To compute the result
result = mean_time.compute()
print(ddf.size)
print(ddf.columns)
print(ddf.shape)
print(ddf.info())
# Perform operations
ddf = ddf.drop_duplicates()
result = ddf.count()

ddf.columns = ['FL_DATE', 'AIRLINE', "AIRLINE_DOT", 'AIRLINE_CODE', 'DOT_CODE', 'FL_NUMBER', 'ORIGIN', 'ORIGIN_CITY',
                      'DEST', 'DEST_CITY', 'CRS_DEP_TIME', 'DEP_TIME', 'DEP_DELAY', 'TAXI_OUT',
                      'WHEELS_OFF', 'WHEELS_ON', 'TAXI_IN', 'CRS_ARR_TIME', 'ARR_TIME', 'ARR_DELAY', 'CANCELLED',
                      'CANCELLATION_CODE', 'DIVERTED', 'CRS_ELAPSED_TIME', 'ELAPSED_TIME', 'AIR_TIME', "DISTANCE",
                      "DELAY_DUE_CARRIER", "DELAY_DUE_WEATHER", "DELAY_DUE_NAS", "DELAY_DUE_SECURITY","DELAY_DUE_LATE_AIRCRAFT"]


null_counts = ddf.isnull().sum()

ddf.isnull().sum()

# Measure time after
end_time = time.time()

# Measure memory usage after
memory_after = psutil.Process().memory_info().rss / (1024 * 1024)  # in megabytes

# Calculate differences
execution_time = end_time - start_time
memory_difference = memory_after - memory_before

# Print results
print(f"Execution Time: {execution_time} seconds")
print(f"Memory Usage Before: {memory_before} MB")
print(f"Memory Usage After: {memory_after} MB")
print(f"Memory Usage for dask in processing data: {memory_difference} MB")

dd.Scalar<size-ag..., dtype=int64>
Index(['FL_DATE', 'AIRLINE', 'AIRLINE_DOT', 'AIRLINE_CODE', 'DOT_CODE',
       'FL_NUMBER', 'ORIGIN', 'ORIGIN_CITY', 'DEST', 'DEST_CITY',
       'CRS_DEP_TIME', 'DEP_TIME', 'DEP_DELAY', 'TAXI_OUT', 'WHEELS_OFF',
       'WHEELS_ON', 'TAXI_IN', 'CRS_ARR_TIME', 'ARR_TIME', 'ARR_DELAY',
       'CANCELLED', 'CANCELLATION_CODE', 'DIVERTED', 'CRS_ELAPSED_TIME',
       'ELAPSED_TIME', 'AIR_TIME', 'DISTANCE', 'DELAY_DUE_CARRIER',
       'DELAY_DUE_WEATHER', 'DELAY_DUE_NAS', 'DELAY_DUE_SECURITY',
       'DELAY_DUE_LATE_AIRCRAFT'],
      dtype='object')
(Delayed('int-af4f193e-218e-48b9-8355-9cb76330ce85'), 32)
<class 'dask.dataframe.core.DataFrame'>
Columns: 32 entries, FL_DATE to DELAY_DUE_LATE_AIRCRAFT
dtypes: object(9), float64(19), int64(4)None
Execution Time: 4.182555198669434 seconds
Memory Usage Before: 4958.61328125 MB
Memory Usage After: 5691.1171875 MB
Memory Usage for dask in processing data: 732.50390625 MB


we can observe that in pandas we used 20+ seconds just for loading the data, while in dask, involving computing time we just need 4.2 seconds, this is how effecient that can be if we utilize all of our laptop cores to process data.

##COMPARATIVE ANLYSIS

#### Processing time and Memory Usage:
| Strategy | Time (in seconds) | Memory Usage (in Mega Byte) | Analysis|      
|----------|------|--------------|---------|
|Traditional(pandas)|22.1|732.4|Pandas use single core for data manipulation, it require more time for processing.|
|Load Less Data|25.27|709.5|If we only load certain data, eliminating some coulumns, definitely the memory usage will be dropped, but if we drop only few coulumns, the processing time will take longer.|
|Optimize Data Types|2.32|452|The time is only counted for the optimizing the data types, not including the loading of dataset, it significantly reduce the memory usage and save time in long run.|
|Sampling (percentage_sample)|0.98|75.5|Uses a percentage-based sample, which reduces memory usage and greatly improves processing time.|
|Parallelize with Dask|4.183|732.5|Utilizes Dask for parallel processing, resulting in a substantial reduction time but same memory usage.|
|Chunking|24.23|1385.9|We see a higher processing time and memory usage because we also perform calculation in chunking, anyway it definitely can save memory usage and processing time becase the dataset is separated and processed with higher speed.|


##Advantages and Disadvantages:
1. Pandas:

-Advantages: Very convenient and commonly used, can find many solutions in internet.

-Disadvantages: Highly time and memory usage consuming especially when dataset is large.

2. Load Less Data:

-Advantages: Save processing time and memory usage, and we able analyze the more important information.

-Disadvantages: Some underlying important data maybe not included in the analysis and causing information loss.

3. Optimize Data Types:

-Advantages: Greatly save memory usage and processing time in long run.

-Disadvantages: Some datatypes maybe went wrong and causing the further analysis not feasible, and if dataset is small this will add some additionaly processing time and add-up the risk.

4. Sampling:

-Advantages: Significantly save processing time and memory usage, the right technique can make the dataset posses certain representability.

-Disadvantages: Some important information maybe lost, especially some rare and important event happened and we failed to capture it.

5. Parallelize with Dask:

-Advantages: Process the dataset parallely and we able to function all the cores of our laptop or computer, and so we can save processing time in long run.

-Disadvantages: Because the cores are not really function that systematically, it will add-up some learning cost in long run.

6. Chunking:

-Advantages: Process the dataset in many separated pieces, then combined to form the results to reduce the memory usage.

-Disadvantages: Processing time will be higher, and we need to decide an appropriate chunking size before we process the data.

##Conclusion

In Summary, in this assignment we found that there are various technique can be used when dealing with a large dataset that are larger than 1 GB in order to avoid cpu crash. Pandas is very useful but since it only utilize one of the cores, it is not fully exploring the capability of the CPU or GPU, in results of high processing time and also the memory usage. Besides that, if there are many unnecessary features in the dataset, we can eliminate them in order to reduce processing time and memory usage. If we want to keep them but at a lower computational cost, we can use sampling based on percentage and the selection is random so that we can select the data distributed in the dataset to make the sample representative of the dataset.We can also optimize the data types so that the data can be stored in a types that required lesser memory usage, and thus we also save processing time in long run and we are not compromised to eliminate any of the data. Last but not least, we can process the data through chunking, so that we separate the dataset and process it with higher efficient, then the processed data is combined to form the results in order to save the memory usage but it will inevitably increase the processing time. We always need to be smart to select the strategies and choose which aspects to compromize in order to achieve our goals.